apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -0,0 +1,247 @@
+            set => _emulatorDetection = GaxPreconditions.CheckEnumValue(value, nameof(value));
+        }
+
+        /// <summary>
+        /// The project ID, or null if this has not yet been configured. Note that a project ID must be configured
+        /// (or detected via an emulator environment variable) before building the <see cref="DatastoreDb"/>.
+        /// </summary>
+        public string ProjectId { get; set; }
+
+        /// <summary>
+        /// Common code for handling the emulator configuration.
+        /// </summary>
+        private ConfiguredBuilder PrepareBuilder()
+        {
+            var clientBuilder = new DatastoreClientBuilder
+            {
+                Endpoint = Endpoint,
+                CallInvoker = CallInvoker,
+                ChannelCredentials = ChannelCredentials,
+                CredentialsPath = CredentialsPath,
+                JsonCredentials = JsonCredentials,
+                Scopes = Scopes,
+                Settings = Settings,
+                TokenAccessMethod = TokenAccessMethod
+            };
+            string projectId = ProjectId;
+
+            if (EmulatorDetection != EmulatorDetection.None)
+            {
+                var emulatorConfig = EmulatorConfiguration.FromEnvironment();
+                switch (EmulatorDetection)
+                {
+                    case EmulatorDetection.ProductionOnlyThrowOnEmulatorEnvironment:
+                        GaxPreconditions.CheckState(
+                            emulatorConfig.ProjectId == null && emulatorConfig.Endpoint == null,
+                            "Emulator environment variables are set, contrary to use of {0}.{1}",
+                            nameof(EmulatorDetection), nameof(EmulatorDetection.ProductionOnlyThrowOnEmulatorEnvironment));
+                        break;
+                    case EmulatorDetection.EmulatorOnly:
+                        GaxPreconditions.CheckState(
+                            emulatorConfig.Endpoint != null,
+                            "Expected {0} environment variable to be set", EmulatorConfiguration.EmulatorHostVariable);
+                        ApplyEmulatorConfiguration();
+                        break;
+                    case EmulatorDetection.ProductionOrEmulator:
+                        if (emulatorConfig.Endpoint != null)
+                        {
+                            ApplyEmulatorConfiguration();
+                        }
+                        else
+                        {
+                            GaxPreconditions.CheckState(emulatorConfig.ProjectId == null, "{0} should not be set when {1} is not", EmulatorConfiguration.EmulatorProjectVariable, EmulatorConfiguration.EmulatorProjectVariable);
+                        }
+                        break;
+                }
+
+                void ApplyEmulatorConfiguration()
+                {
+                    // TODO: Do we want to require this? Or should a configured emulator override any credentials?
```

Comment:
I'd say we don't want to require this, so that client code can stay the same in any environment. If there's an emulator configured and we should use it, then, we use that overriding.